File: Activity Selection.java
class Solution {
    class Pair{
        int start;
        int finish;
        public Pair(int start, int finish){
            this.start = start;
            this.finish = finish;
        }
    }
    public int activitySelection(int[] start, int[] finish) {
        PriorityQueue<Pair> pq = new PriorityQueue<>((a, b) -> a.finish != b.finish ? Integer.compare(a.finish, b.finish) : Integer.compare(a.start, b.start));
        for(int i = 0; i < start.length; i++){
            pq.offer(new Pair(start[i], finish[i]));
        }
        
        int count = 0;
        int time = 0;
        while(!pq.isEmpty()){
            Pair p = pq.poll();
            if(p.start > time){
                count++;
                time = p.finish;
            }
        }
        
        return count;
    }
}

File: Activity Selection.cpp
class Solution {
    struct Pair {
        int start;
        int finish;
        Pair(int s, int f) : start(s), finish(f) {}
    };

    struct cmp {
        bool operator()(const Pair& a, const Pair& b) const {
            if (a.finish != b.finish)
                return a.finish > b.finish;
            return a.start > b.start;
        }
    };

public:
    int activitySelection(vector<int>& start, vector<int>& finish) {
        priority_queue<Pair, vector<Pair>, cmp> pq;
        for (size_t i = 0; i < start.size(); i++) {
            pq.push(Pair(start[i], finish[i]));
        }

        int count = 0;
        int time = 0;
        while (!pq.empty()) {
            Pair p = pq.top();
            pq.pop();
            if (p.start > time) {
                count++;
                time = p.finish;
            }
        }
        return count;
    }
};
